File: baekjoon_all/07000+/boj_7579.cpp
// Solve 2023-01-05
// Update 2024-03-27

#include <bits/stdc++.h>
using namespace std;

#define FASTIO ios_base::sync_with_stdio(false);cin.tie(NULL); // boj_15552.cpp
#define SETPRECISION(n) cout << fixed;cout.precision(n); // boj_1008.cpp
#define SIZE(v) (int)v.size()
#define ALL(v) v.begin(),v.end()
using ll = long long;

int main() {
    FASTIO;

    int n, m;
    cin >> n >> m;

    int memory[100];

    for (int i = 0; i < n; i++) {
        cin >> memory[i];
    }

    int cost[100];

    for (int i = 0; i < n; i++) {
        cin >> cost[i];
    }

    // memo[c] : c의 비용으로 확보할 수 있는 최대한의 메모리
    int memo[10001] = { 0 };

    for (int i = 0; i < n; i++) {
        for (int j = 10000; j >= cost[i]; j--) {
            memo[j] = max(memo[j], memo[j - cost[i]] + memory[i]);
        }
    }

    int idx = lower_bound(memo, memo + 10001, m) - memo;
    cout << idx << '\n';

    return 0;
}

File: baekjoon_all/07000+/boj_7579.java
// Solve 2024-03-27

// 백준에 제출할 때는 class 이름을 Main으로 설정해야 한다.

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class boj_7579 {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        int[] memory = new int[n];
        int[] cost = new int[n];

        st = new StringTokenizer(br.readLine(), " ");

        for (int i = 0; i < n; i++) {
            memory[i] = Integer.parseInt(st.nextToken());
        }

        st = new StringTokenizer(br.readLine(), " ");

        for (int i = 0; i < n; i++) {
            cost[i] = Integer.parseInt(st.nextToken());
        }

        int[] memo = new int[10001];

        for (int i = 0; i < n; i++) {
            for (int j = 10000; j >= cost[i]; j--) {
                memo[j] = Math.max(memo[j], memo[j - cost[i]] + memory[i]);
            }
        }

        int idx = lowerBound(memo, 10001, m);
        System.out.println(idx);
        br.close();
    }

    static int lowerBound(int[] arr, int arrSize, int keyValue) {
        int low = -1;
        int high = arrSize;

        while (low + 1 < high) {
            int mid = (low + high) / 2;

            if (arr[mid] >= keyValue) {
                high = mid;
            } else {
                low = mid;
            }
        }

        return high;
    }

}
